compiler_gym/datasets/files_dataset.py

```diff
@@ -0,0 +1,156 @@
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import os
+from pathlib import Path
+from typing import Iterable, List, Optional
+
+from compiler_gym.datasets.dataset import Benchmark, Dataset
+from compiler_gym.util.decorators import memoized_property
+
+
+class FilesDataset(Dataset):
+    """A dataset comprising a directory tree files.
+
+    A FilesDataset is a root directory that contains (possibly nested) files,
+    where each file represents a benchmark. Files can be filtered on their
+    expected filename suffix.
+
+    Every file that matches the expected suffix is a benchmark. The URI of the
+    benchmarks is the relative path of the file within the dataset, stripped of
+    any matching file suffix. For example, given a dataset root around the
+    directory tree :code:`/tmp/dataset` and with file suffix :code:`.txt`:
+
+    .. code-block::
+
+        /tmp/dataset/a.txt
+        /tmp/dataset/LICENSE
+        /tmp/dataset/subdir/subdir/b.txt
+        /tmp/dataset/subdir/subdir/c.txt
+
+    a FilesDataset representing this directory tree will contain the following
+    URIs:
+
+        >>> list(dataset.benchmark_uris())
+        [
+            "benchamrk://ds-v0/a",
+            "benchamrk://ds-v0/subdir/subdir/b",
+            "benchamrk://ds-v0/subdir/subdir/c",
+        ]
+    """
+
+    def __init__(
+        self,
+        dataset_root: Path,
+        benchmark_file_suffix: str = "",
+        memoize_uris: bool = True,
+        **dataset_args,
+    ):
+        """Constructor.
+
+        :param dataset_root: The root directory to look for benchmark files.
+
+        :param benchmark_file_suffix: A file extension that must be matched for
+            a file to be used as a benchmark.
+
+        :param memoize_uris: Whether to memoize the list of URIs contained in
+            the dataset. Memoizing the URIs is a tradeoff between *O(n)*
+            computation complexity of random access vs *O(n)* space complexity
+            of memoizing the URI list.
+
+        :param dataset_args: See :meth:`Dataset.__init__()
+            <compiler_gym.datasets.Dataset.__init__>`.
+        """
+        super().__init__(**dataset_args)
+        self.dataset_root = dataset_root
+        self.benchmark_file_suffix = benchmark_file_suffix
+        self.memoize_uris = memoize_uris
+        self._memoized_uris = None
+
+    @memoized_property
+    def n(self) -> int:  # pylint: disable=invalid-overriden-method
+        self.install()
+        return sum(
+            sum(1 for f in files if f.endswith(self.benchmark_file_suffix))
+            for (_, _, files) in os.walk(self.dataset_root)
+        )
+
+    @property
+    def _benchmark_uris_iter(self) -> Iterable[str]:
+        """Return an iterator over benchmark URIs is consistent across runs."""
+        self.install()
+        for root, dirs, files in os.walk(self.dataset_root):
+            dirs.sort()
```

Comment:
what does the sort do here?

Reply:
It ensures that the order of this iterator is consistent across runs. [os.walk() calls os.scandir()](https://github.com/python/cpython/blob/a07da09ad5bd7d234ccd084a3a0933c290d1b592/Lib/os.py#L357), and [os.scandir() iteration order is arbitrary](https://docs.python.org/3/library/os.html#os.scandir).

Reply:
I'll add a comment to clarify.